federation/planner.go

```diff
@@ -0,0 +1,300 @@
+package federation
+
+import (
+	"errors"
+	"fmt"
+	"sort"
+
+	"github.com/samsarahq/thunder/graphql"
+)
+
+type StepKind int
+
+const (
+	KindType StepKind = iota
+	KindField
+)
+
+type PathStep struct {
+	Kind StepKind
+	Name string
+}
+
+type Plan struct {
+	Path    []PathStep
+	Service string
+	Kind    string
+	// XXX: What are we using Type for here again? -- oh, it's for the __federation field...
+	Type         string
+	SelectionSet *graphql.SelectionSet
+	After        []*Plan
+}
+
+type Executor struct {
+	schema    *SchemaWithFederationInfo
+	flattener *flattener
+}
+
+
+func (e *Executor) planObject(typ *graphql.Object, selectionSet *graphql.SelectionSet, service string) (*Plan, error) {
+	p := &Plan{
+		Type:         typ.Name,
+		Service:      service,
+		SelectionSet: &graphql.SelectionSet{},
+		After:        nil,
+		Kind:         "query",
+	}
+
+	// XXX: pass in sub-path (and sub-plan slice?) to make sub-plan munging simpler?
+	// - should we type check here?
+
+	// - refactor to return array of subplans with preferential treatment for given service?
+	// eh whatever.
+
+	var localSelections []*graphql.Selection
+	selectionsByService := make(map[string][]*graphql.Selection)
+
+	// Assert that we are working with a flattened query.
+	if len(selectionSet.Fragments) > 0 {
+		return nil, errors.New("selectionSet has fragments, expected flattened query")
+	}
+
+	for _, selection := range selectionSet.Selections {
+		if selection.Name == "__typename" {
+			localSelections = append(localSelections, selection)
+			continue
+		}
+
+		field, ok := typ.Fields[selection.Name]
+		if !ok {
+			return nil, fmt.Errorf("typ %s has no field %s", typ.Name, selection.Name)
+		}
+
+		fieldInfo := e.schema.Fields[field]
+
+		// if we can stick to the current service, stay there
+		if fieldInfo.Services[service] {
+			localSelections = append(localSelections, selection)
+		} else {
+			serviceWithField := ""
+
+			for service, hasField := range fieldInfo.Services {
+					if hasField {
+						serviceWithField = service
+					}
+			}
+
+			selectionsByService[serviceWithField] = append(
+				selectionsByService[serviceWithField], selection)
+		}
+	}
+
+	for _, selection := range localSelections {
+		// we have already checked above that this field exists
+		field := typ.Fields[selection.Name]
+		var childPlan *Plan
+		if selection.SelectionSet != nil {
+			// XXX: assert existence of types elsewhere?
+			var err error
+			// XXX type assertoin
+			childPlan, err = e.plan(field.Type, selection.SelectionSet, service)
+			if err != nil {
+				return nil, fmt.Errorf("planning for %s: %v", selection.Name, err)
+			}
+		}
+
+		// newSelection := &graphql.RawSelection{
```

Comment:
can we make a copy here still? i don't want to modify the incoming query if we can help it.